mona_openai/mona_openai_create.py

```diff
@@ -0,0 +1,132 @@
+import time
+from collections.abc import Callable
+from .util.async_util import (
+    call_non_blocking_sync_or_async
+)
+from .util.func_util import add_conditional_sampling
+from .util.openai_util import get_model_param
+from .util.stream_util import ResponseGatheringIterator
+from .util.tokens_util import get_usage
+from .util.object_util import get_subscriptable_obj
+
+MONA_ARGS_PREFIX = "MONA_"
+CONTEXT_ID_ARG_NAME = MONA_ARGS_PREFIX + "context_id"
+EXPORT_TIMESTAMP_ARG_NAME = MONA_ARGS_PREFIX + "export_timestamp"
+ADDITIONAL_DATA_ARG_NAME = MONA_ARGS_PREFIX + "additional_data"
+
+
+
+
+async def create_logic(
+        is_async,
+        export_function: Callable,
+        super_function: Callable,
+        logging_message_getter,
+        sampling_ratio,
+        all_prompt_texts_getter,
+        all_response_texts_getter,
+        stream_delta_text_from_choice_getter,
+        final_choice_getter,
+        specs,
+        args,
+        kwargs):
+    is_stream = kwargs.get("stream", False)
+
+    response = None
+
+    # will be used only when stream is enabled
+    stream_start_time = None
+
+    async def _inner_log_message(is_exception):
+        subscriptable_response = get_subscriptable_obj(response)
+        
+        return await call_non_blocking_sync_or_async(
+            export_function,
+            (
+                logging_message_getter(
+                    kwargs,
+                    start_time,
+                    is_exception,
+                    is_async,
+                    stream_start_time,
+                    subscriptable_response,
+                ),
+                kwargs.get(
+                    CONTEXT_ID_ARG_NAME,
+                    subscriptable_response["id"] if subscriptable_response else None,
+                ),
+                kwargs.get(EXPORT_TIMESTAMP_ARG_NAME, start_time),
+            ),
+        )
+
+    log_message = add_conditional_sampling(
+        _inner_log_message, sampling_ratio
+    )
+
+    start_time = time.time()
+
+    async def inner_super_function():
+        # Call the actual openai create function without the Mona
+        # specific arguments.
+        return await call_non_blocking_sync_or_async(
+            super_function,
+            args,
+            {
+                x: kwargs[x]
+                for x in kwargs
+                if not x.startswith(MONA_ARGS_PREFIX)
+            },
+        )
+
+    async def inner_handle_exception():
+        if not specs.get("avoid_monitoring_exceptions", False):
+            await log_message(True)
+
+    if not is_stream:
+        try:
+            response = await inner_super_function()
+        except Exception:
+            await inner_handle_exception()
+            raise
+
+        await log_message(False)
+
+        return response
+
+    # From here it's stream handling.
```

Comment:
nit: I don't love this comment here, if we add code below this point it might not age so well... Perhaps make these into a comment about the _stread_done_callback and another one for before the "try" below such that they'll only apply to the code block afterwards?

Reply:
We don't really need this comment at all

Reply:
The if above is clear enough